tokio/src/task/task_local.rs

```diff
@@ -161,34 +175,85 @@ impl<T: 'static> LocalKey<T> {
     /// });
     /// # }
     /// ```
+    ///
+    /// [`with`]: fn@Self::with
+    /// [`try_with`]: fn@Self::try_with
+    #[track_caller]
     pub fn sync_scope<F, R>(&'static self, value: T, f: F) -> R
     where
         F: FnOnce() -> R,
     {
-        let scope = TaskLocalFuture {
-            local: self,
-            slot: Some(value),
-            future: (),
-            _pinned: PhantomPinned,
-        };
-        crate::pin!(scope);
-        scope.with_task(|_| f())
+        let mut value = Some(value);
+        match self.scope_inner(&mut value, f) {
+            Ok(res) => res,
+            Err(ScopeInnerErr::BorrowError) => {
+                panic!("sync_scope called while Task Local Storage is borrowed")
+            }
+            Err(ScopeInnerErr::AccessError) => {
+                panic!("cannot access a Task Local Storage value during or after destruction")
+            }
+        }
+    }
+
+    fn scope_inner<F, R>(&'static self, slot: &mut Option<T>, f: F) -> Result<R, ScopeInnerErr>
+    where
+        F: FnOnce() -> R,
+    {
+        struct Guard<'a, T: 'static> {
+            local: &'static LocalKey<T>,
+            slot: &'a mut Option<T>,
+        }
+
+        impl<'a, T: 'static> Drop for Guard<'a, T> {
+            fn drop(&mut self) {
+                // This should not panic.
+                //
+                // We know that the RefCell was not borrowed before the call to
+                // `scope_inner`, so the only way for this to panic is if the
+                // closure has created but not destroyed a RefCell guard.
+                // However, we never give user-code access to the guards, so
+                // there's no way for user-code to forget to destroy a guard.
+                //
+                // The call to `with` also should not panic, since the
+                // thread-local wasn't destroyed when we first called
+                // `scope_inner`, and it shouldn't have gotten destroyed since
+                // then.
+                self.local.inner.with(|inner| {
+                    let mut ref_mut = inner.borrow_mut();
+                    std::mem::swap(self.slot, &mut *ref_mut);
```

Comment:
tiny style nit: we use multiple `std::mem` functions in this file (`swap` and `needs_drop`), i might consider a `use std::mem;`